### PR TITLE
[COM-1558] Add Segmented Button

### PR DESCRIPTION
**Pull request by Claratj**
**Checklist**

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Please check if the PR fulfills these requirements**

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Storybook or docs have been added / updated (for bug fixes / features)

**What is the ticket / issue associated with this change?** (Link to the ticket or issue)

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

**Storybook Link** ([Link to a local storybook url to review new functionality](https://5eed577398d08d0022b52d76-tystwhpekx.chromatic.com/))

### CODE REVIEW

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      {trailingIconSized}
+    </ChakraButton>
+  );
+};
+
+SegmentedButton.Button = Button;
+
+const IconButton: FC<SegmentedIconButtonProps> = ({ isActive, icon, size, variant, ...rest }) => {
+  const styles = useMultiStyleConfig('SegmentedButton', { isActive, variant, size }) as SegmentedButtonStyles;
+  const iconChild = getIcon(icon, size);
```
Comment (guilean) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx:66]: Rename to `iconSized` to be consistent with the Button implementation 🙏🏻

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+const Template = (props) => <SegmentedButton {...props} key={props.defaultOption} options={optionsButton} />;
+
+export const Playground = Template.bind({});
+Playground.args = {
+  variant: 'primary',
+  size: 's',
+  isDisabled: false,
+  defaultOption: '1',
+};
```
Comment (guilean) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:114]: Could you add the `onChange` to the Template too? 🙏🏻

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx (hunk 1 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx]: this has been done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx (hunk 2 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx]: added!

**Review by guilean — APPROVED** [review]

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+              <Td>{state}</Td>
+              {segmentedButtonSizes.map((size, i) => (
+                <Td key={`${variant}-${state}-${size}-${i}-icon`}>
+                  <Flex flexDirection="column" alignItems="start" gap="5">
+                    <SegmentedButton
+                      size={size}
+                      variant={variant}
+                      options={optionsIcon}
+                      defaultOption={'facebook'}
+                      onChange={(value) => console.log(value)}
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:82]: nit:
```suggestion
                      onChange={console.log}
```

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+              <Td>{variant}</Td>
+              <Td>{state}</Td>
+              {segmentedButtonSizes.map((size, i) => (
+                <Td key={`${variant}-${state}-${size}-${i}-icon`}>
+                  <Flex flexDirection="column" alignItems="start" gap="5">
+                    <SegmentedButton
+                      size={size}
+                      variant={variant}
+                      options={optionsIcon}
+                      defaultOption={'facebook'}
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:81]: nit: 
```suggestion
                      defaultOption="facebook"
```

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                      onChange={(value) => console.log(value)}
+                      {...(state === 'Disabled' && { isDisabled: true })}
+                    />
+
+                    <SegmentedButton
+                      size={size}
+                      variant={variant}
+                      options={optionsButton}
+                      defaultOption={'1'}
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:90]: nit:
```suggestion
                      defaultOption="1"
```

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                      onChange={(value) => console.log(value)}
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:91]:
```suggestion
                      onChange={console.log}
```

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  const options = [
+    { value: 'dummy_value_1', segment: <SegmentedButton.Button>dummy text 1</SegmentedButton.Button> },
+    { value: 'dummy_value_2', segment: <SegmentedButton.Button>dummy text 2</SegmentedButton.Button> },
+    { value: 'dummy_value_3', segment: <SegmentedButton.Button>dummy text 3</SegmentedButton.Button> },
+  ];
+  const onChangeMock = jest.fn();
+  const givenComponentRendered = (props?: Partial<SegmentedButtonProps>) =>
+    renderWithProviders(<SegmentedButton options={options} onChange={onChangeMock} {...props} />);
+
+  it('should render without crashing', () => {
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx:16]: nit
```suggestion
  it('should render', () => {
```

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    givenComponentRendered();
+    expect(screen.getByText('dummy text 1')).toBeInTheDocument();
+  });
+  it('should call onChange when a segment is clicked', () => {
+    givenComponentRendered();
+    const dummy2Button = screen.getByText('dummy text 2');
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx:22]:
```suggestion
    const dummy2Button = screen.getByRole('button', { name: 'dummy text 2' });
```

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    fireEvent.click(dummy2Button);
+    expect(onChangeMock).toHaveBeenCalledWith('dummy_value_2');
+  });
+  it('should disable the segments when is disabled', () => {
+    givenComponentRendered({ isDisabled: true });
+    expect(screen.getByText('dummy text 1')).toBeDisabled();
+  });
+
+  it('should select an option by default when defaultOption is provided', () => {
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx:31]: is it possible to verify that the selected button is the one associated to the given value?

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  ButtonProps,
+  OptionValue,
+  SegmentedButtonProps,
+  SegmentedButtonSize,
+  SegmentedButtonStaticMembers,
+  SegmentedButtonStyles,
+  SegmentedIconButtonProps,
+} from './types';
+
+export const SegmentedButton: FC<SegmentedButtonProps> & SegmentedButtonStaticMembers = ({
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx:15]: is it possible to export this as a forwardReference? Not a _hard requirement_ but definitely a nice to have

File: packages/components/src/components/form/SegmentedButton/index.tsx [hunk 11]
```diff
@@ -0,0 +1 @@
+export * from './SegmentedButton';
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/index.tsx:1]: we should export the types too, at least the one _relevant_ to the component usage

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:1]: the way this story is written is not helpful to illustrate the component usage in the storybook docs, can we implement at least on the the story in a way that shows how the component should be used from the code perspective please?

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx (hunk 4 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:81]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx (hunk 3 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:82]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx (hunk 5 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:90]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx (hunk 6 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:91]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx (hunk 7 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx:16]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx (hunk 8 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx:22]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx (hunk 9 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.test.tsx:32]: We added a data-value to verify it. Would it be enough?

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx (hunk 10 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx:15]: Done it

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.stories.tsx:1]: Changed, tell me if what I have modified would be enough or would we need something else.

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+      {(children as ReactElement[]).map((child) =>
+        cloneElement(child, {
+          key: child.props.value,
+          onClick: () => handleChange(child.props.value),
+          isActive: selectedValue === child.props.value,
+          variant,
+          size: responsiveSize,
+          isDisabled,
+          'aria-current': selectedValue === child.props.value,
+          'data-value': child.props.value,
```
Comment (guilean) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx:43]: @Claratj given that we are now using the `value` property I think we could omit the `data-value`. You'll need to update the tests too

Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+          'aria-current': selectedValue === child.props.value,
+        })
+      )}
+    </Flex>
+  );
+};
+
+const Button = forwardRef<SegmentedButtonButtonProps, typeof ChakraButton>(
+  ({ isActive, children, leadingIcon, trailingIcon, size, variant, ...rest }, ref) => {
+    const styles = useMultiStyleConfig('SegmentedButton', { isActive, variant, size }) as SegmentedButtonStyles;
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx:51]: I have mixed feeling with calling this in all children, I tend to remember that in this cases chakra-ui implements a Context and pass the styles down and we are already doing something similar in another component

File: packages/components/src/components/form/SegmentedButton/types.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
+import { ButtonProps as ChakraButtonProps, ResponsiveValue, StyleProps } from '@chakra-ui/react';
+import { FC, ReactElement } from 'react';
+
+export const segmentedButtonVariants = ['primary', 'primary-alt', 'secondary', 'secondary-alt'] as const;
+export type SegmentedButtonVariant = (typeof segmentedButtonVariants)[number];
+export const segmentedButtonSizes = ['xs', 's', 'm', 'l'] as const;
+export type SegmentedButtonSize = (typeof segmentedButtonSizes)[number];
+
+export type Value = string | number;
```
Comment (tkanzakic) [inline packages/components/src/components/form/SegmentedButton/types.ts:9]: this type should be scoped to the component, `Value` too generic to be exported out
```suggestion
export type SegmentedButtonValue = string | number;
```

Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/types.ts:9]: done!

File: packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx (hunk 13 above)
Comment (Claratj) [reply packages/components/src/components/form/SegmentedButton/SegmentedButton.tsx]: done!

**Review by tkanzakic — APPROVED** [review]

**Pull request merged**